src/npg_irods/metadata/common.py

```
@@ -98,6 +104,169 @@ class CompressSuffix(AsValueEnum):
+        raise ChecksumError(
+            f"Found {len(checksums)} checksums in iRODS metadata",
+            path=obj,
+            expected=obj.checksum(),
+            observed=checksums,
+        )
+
+    if not has_matching_checksums(obj):
+        return False
+
+    if not has_checksum_metadata(obj):
+        return False
+
+    return AVU(DataFile.MD5, obj.checksum()) in checksum_meta
+
+
+def ensure_matching_checksum_metadata(obj: DataObject) -> bool:
+    """Ensure that a data object has checksum metadata that matches its iRODS checksums.
+
+    This check implies that the data object has full checksum coverage of all its
+    valid replicas, and that those checksums all match each other. This function first
+    attempts to ensure that this is the case. If this cannot be achieved, a
+    ChecksumError is raised.
+
+    If there is no checksum metadata on the data object, a new AVU is added and this
+    function returns True. If there is already checksum metadata on the data object
+    that concurs with the current checksum, this function does nothing and returns
+    False.
+
+    If there is already checksum metadata that does not concur with the
+    current checksum, a ChecksumError is raised.
+
+    Args:
+        obj: The data object to repair.
+
+    Returns:
+        True if a repair was done.
+    """
+    if has_matching_checksum_metadata(obj):
+        return False
+
+    valid_replicas = [r for r in obj.replicas() if r.valid]
+    expected_checksums = [obj.checksum()] * len(valid_replicas)
+    observed_checksums = [r.checksum for r in valid_replicas]
+
+    if not has_complete_checksums(obj):
+        raise ChecksumError(
+            "Failed to ensure that the data object has matching checksum metadata "
+            "because not all of its valid replicas have a checksum",
+            path=obj,
+            expected=expected_checksums,
+            observed=observed_checksums,
+        )
+
+    if not has_matching_checksums(obj):
```

You are running has_complete_checksums twice, wouldn't it be better to raise the exception inside the functions, which gives you the same information and means you only need to call one of them?

---

I'd rather have predicates that don't raise exceptions around the thing they're testing (so we get `True`/`False` vs. `True`/`<Exception>`) and leave the caller to decide whether the exception is necessary. Yes, it's called twice - if that becomes a bottleneck, I would put a cache decorator on it.